### ISSUE TITLE
Jimple Walker fails on invocations when the target method's declaring class is dangling

### ISSUE THREAD

**Issue by ensoftkevin** [soot-oss/soot#369]
Issue #298 is related to this issue, though I think the problem existed even before then, it was just better hidden.  That issue introduced code in JInterfaceInvokeExpr and JVirtualInvokeExpr to check if the declaring class of the target method is resolved to at least HIERARCHY level.  This is presumably so the call to `methodRef.declaredClass().isInterface()` is guaranteed to return an accurate result, all with the end goal of throwing an exception if the wrong type of `InvokeExpr` was chosen by the user.  However, because the code calls `checkLevelWithoutResolving` instead of `checkLevel`, the check will fail even if the Scene hasn't finished resolving yet.  Based on code comments, this appears to be by design.

This is all problematic for the Jimple `Walker`, because it still knows the proper expression type for an invocation even when the target method's declaring class is still dangling.  It knows this because it is given a parser `Node`, such as `AVirtualNonstaticInvoke` or `AInterfaceNonstaticInvoke`, and it simply wants to construct a matching `InvokeExpr`.  But if the target method's declaring class is dangling, the thrown exception during level checking prevents it from succeeding.

I'm not sure how it would be best implemented, but there should be an 'unsafe' way to construct method invocation expressions, so the walker can get around this problem.

**Comment by ericbodden** [comment]
@ensoftkevin under which circumstances is this ever a problem. Do you have a test case in which you see the respective exception being thrown? If so, please share it.

**Comment by ensoftkevin** [comment]
I don't know if it can be reproduced from Soot proper.  I looked at the code, and it seems that when resolving a Scene, Soot takes some care to ensure that referenced types are resolved to HIERARCHY level before attempting to resolve any method bodies.

However, we are using Soot as a library in our project.  In our case, we don't want to bother with resolving the entire scene, we just want to parse and load a single SootClass in isolation, and we don't care if references in the class are left dangling.  So we are using `JimpleAST.createSootClass()` to do this, which internally uses a `Walker` to process the parser nodes into Soot values.

But as previously described, we found an exception is thrown when the walker attempts to construct `JInterfaceInvokeExpr` or `JVirtualInvokeExpr` expressions.  This is because these constructors are coded to check if the declaring class for the method ref is resolved to HIERARCHY level.  Of course, in our case, they are only resolved to DANGLING level.

From what I can tell, the resolving level check is only used to guarantee that `methodRef.declaringClass().isInterface()` will return a valid result.  And that call is only used to verify that the expression being constructed is compatible with the provided method reference.  However, with the `Walker`, that check is unnecessary...for instance if the source is an `AVirtualNonstaticInvoke` node, then that is evidence enough that `JVirtualInvokeExpr` is the correct Soot node to construct.
